src/dd/StateGeneration.cpp

```diff
@@ -62,8 +65,208 @@ vCachedEdge randomNode(Qubit v, vNode* left, vNode* right, Generator& gen,
+ * recursive algorithm.
+ *
+ * @param begin Iterator pointing to the beginning of the state vector.
+ * @param end Iterator pointing to the end of the state vector.
+ * @param v The current level of recursion. Starts at the highest level of
+ * the state vector (log base 2 of the vector size - 1).
+ * @param dd The DD package to use.
+ * @return A vCachedEdge representing the root node of the created DD.
+ *
+ * @details This function recursively breaks down the state vector into halves
+ * until each half has only one element. At each level of recursion, two new
+ * edges are created, one for each half of the state vector. The two resulting
+ * decision diagram edges are used to create a new decision diagram node at
+ * the current level, and this node is returned as the result of the current
+ * recursive call. At the base case of recursion, the state vector has only
+ * two elements, which are converted into terminal nodes of the decision
+ * diagram.
+ *
+ * @note This function assumes that the state vector size is a power of two.
+ */
+vCachedEdge makeStateFromVector(const CVec::const_iterator& begin,
+                                const CVec::const_iterator& end, const Qubit v,
+                                Package& dd) {
+  if (v == 0U) {
+    const auto zeroSuccessor = vCachedEdge::terminal(*begin);
+    const auto oneSuccessor = vCachedEdge::terminal(*(begin + 1));
+    return dd.makeDDNode<vNode, CachedEdge>(0, {zeroSuccessor, oneSuccessor});
+  }
+
+  const auto pivot = std::next(begin, std::distance(begin, end) / 2);
+  const auto zeroSuccessor = makeStateFromVector(begin, pivot, v - 1, dd);
+  const auto oneSuccessor = makeStateFromVector(pivot, end, v - 1, dd);
+  return dd.makeDDNode<vNode, CachedEdge>(v, {zeroSuccessor, oneSuccessor});
+}
+
 } // namespace
 
+VectorDD makeZeroState(const std::size_t n, Package& dd,
+                       const std::size_t start) {
+  const std::vector<BasisStates> state(n - start, BasisStates::zero);
+  return makeBasisState(n, state, dd, start);
+}
+
+VectorDD makeBasisState(const std::size_t n, const std::vector<bool>& state,
+                        Package& dd, const std::size_t start) {
+  const auto op = [](bool b) {
+    return b ? BasisStates::one : BasisStates::zero;
+  };
+  std::vector<BasisStates> bState(state.size());
+  std::transform(state.begin(), state.end(), bState.begin(), op);
+  return makeBasisState(n, bState, dd, start);
+}
+
+VectorDD makeBasisState(const std::size_t n,
+                        const std::vector<BasisStates>& state, Package& dd,
+                        const std::size_t start) {
+  suitablePackage(n + start, dd);
+
+  if (state.size() < n) {
+    throw std::invalid_argument(
+        "Insufficient qubit states provided. Requested " + std::to_string(n) +
+        ", but received " + std::to_string(state.size()));
+  }
+
+  vCachedEdge f = vCachedEdge::one();
+  for (std::size_t p = start; p < n + start; ++p) {
+    switch (state[p]) {
```

Comment:
@burgholzer Am I stupid right now - or shouldn't the loop iterate over `[0, n]` and only the qubit value for the new nodes should be `p + start`?

Given the safety check above (`size() < n`) this should be an out-of-bounds for any `start > 0`, right?